private/bufpkg/bufmodule/stats.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufmodule
+
+import (
+	"context"
+
+	"github.com/bufbuild/protocompile/ast"
+	"github.com/bufbuild/protocompile/parser"
+	"github.com/bufbuild/protocompile/reporter"
+	"go.uber.org/multierr"
+)
+
+// ComputeStats gathers some simple stats about the size of the given module.
+func ComputeStats(ctx context.Context, module Module) (*Stats, error) {
+	infos, err := module.SourceFileInfos(ctx)
+	if err != nil {
+		return nil, err
+	}
+	handler := reporter.NewHandler(reporter.NewReporter(
+		func(reporter.ErrorWithPos) error {
+			// never aborts
+			return nil
+		},
+		nil,
+	))
+	var errs error
+	stats := &Stats{}
+	for _, info := range infos {
+		file, err := module.GetModuleFile(ctx, info.Path())
+		if err != nil {
+			errs = multierr.Append(errs, err)
+			continue
+		}
+		err = func() error {
+			defer func() {
+				errs = multierr.Append(errs, file.Close())
+			}()
+			// This can return an error and non-nil AST.
+			astRoot, err := parser.Parse(info.Path(), file, handler)
+			if astRoot == nil {
+				// No AST implies an I/O error trying to read the
+				// file contents. No stats to collect.
+				return err
+			}
+			if err != nil {
+				// There was a syntax error, but we still have a partial
+				// AST we can examine.
+				stats.NumFilesWithSyntaxErrors++
+			}
+			examineFile(astRoot, stats)
+			return nil
+		}()
+		if err != nil {
+			errs = multierr.Append(errs, err)
+		}
+	}
+	if stats.NumFiles == 0 && errs != nil {
+		return nil, errs
+	}
+	return stats, errs
+}
+
+// Stats represents some statistics/metrics about a module.
+type Stats struct {
```

Comment:
Should this also include the number of packages?

I found this an interesting (and potentially) useful metric, example:

```
owner     module      files  packages  services  methods  messages  enums
beta      googleapis  3149   305       823       4108     14328     2633
beta      envoy       560    365       52        91       1868      218
bufbuild  buf         41     8         34        166      463       27
bufbuild  eliza       1      1         1         3        6         0
```

Reply:
Done